unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs

```diff
@@ -0,0 +1,82 @@
+using TMPro;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace AvatarNamesHUD
+{
+    public class AvatarNamesTracker
+    {
+        private static readonly Vector3 OFFSET = new Vector3(0, 3f, 0);
+        private static readonly int VOICE_CHAT_ANIMATOR_TALKING = Animator.StringToHash("Talking");
+        private const float NAME_VANISHING_POINT_DISTANCE = 15.0f;
+
+        internal readonly RectTransform canvasRect;
+        internal readonly Image background;
+        internal readonly CanvasGroup backgroundCanvasGroup;
+        internal readonly TextMeshProUGUI name;
+        internal readonly CanvasGroup voiceChatCanvasGroup;
+        internal readonly Animator voiceChatAnimator;
+
+        internal PlayerStatus playerStatus;
+        internal bool visibility = false;
+
+        public AvatarNamesTracker(RectTransform canvasRect, RectTransform backgroundRect, RectTransform nameRect, RectTransform voiceChatRect)
+        {
+            this.canvasRect = canvasRect;
+            backgroundCanvasGroup = backgroundRect.GetComponent<CanvasGroup>();
+            background = backgroundRect.GetComponent<Image>();
+            name = nameRect.GetComponent<TextMeshProUGUI>();
+            voiceChatCanvasGroup = voiceChatRect.GetComponent<CanvasGroup>();
+            voiceChatAnimator = voiceChatRect.GetComponent<Animator>();
+        }
+
+        public void SetVisibility(bool visible)
+        {
+            visibility = visible;
+            background?.gameObject.SetActive(visibility);
+            name?.gameObject.SetActive(visibility);
+            voiceChatCanvasGroup?.gameObject.SetActive(visibility && (playerStatus?.isTalking ?? false));
+        }
+
+        public void SetPlayerStatus(PlayerStatus newPlayerStatus)
+        {
+            playerStatus = newPlayerStatus;
+            if (playerStatus == null)
+                return;
+
+            name.text = newPlayerStatus.name;
+            name.ForceMeshUpdate(); //To get the new bounds
+            UpdatePosition();
+        }
+
+        public void UpdatePosition()
+        {
+            if (playerStatus == null)
+                return;
+
+            var mainCamera = Camera.main;
+            Vector3 screenPoint = mainCamera == null ? Vector3.zero : mainCamera.WorldToViewportPoint(playerStatus.worldPosition + OFFSET);
```

Comment:
Remember that when many avatars are present, and avatars are on your back, names for avatars that are outside the viewport can be rendered. We should try to deactivate those renderers.

Reply:
They will be placed outside of the renderer and wont be visible. In any case I will deactivate them to gain that edge.